File: mobile-flutter/lib/data/models/report.dart
class Report {
  final String id;
  final String title;
  final String? category;
  final String? subcategory;
  final String description;
  final String? address;
  final double lat;
  final double lng;
  final String status;
  final String severity;
  final double? severityConfidence;
  final String? severitySource;
  final String? severityReasoning;
  final int similarReportCount;
  final DateTime submittedDate;
  final List<String> photos;
  final List<String> resolutionPhotos;
  final List<String> barangayPhotos;
  final List<StatusEvent> history;
  final String? resolutionDescription;
  final DateTime? resolutionDate;
  final String? barangay;
  final String? reporterName;
  final String? invalidReason;
  final bool barangayConfirmed;
  final String? barangayNote;
  final bool priorityEndorsed;
  final String? priorityEndorsementReason;
  final bool attentionRequested;
  final String? attentionRequestedReason;
  final bool recurringFlagged;

  Report({
    required this.id,
    required this.title,
    this.category,
    this.subcategory,
    required this.description,
    this.address,
    required this.lat,
    required this.lng,
    required this.status,
    required this.severity,
    this.severityConfidence,
    this.severitySource,
    this.severityReasoning,
    required this.similarReportCount,
    required this.submittedDate,
    required this.photos,
    required this.resolutionPhotos,
    required this.barangayPhotos,
    required this.history,
    this.resolutionDescription,
    this.resolutionDate,
    this.barangay,
    this.reporterName,
    this.invalidReason,
    required this.barangayConfirmed,
    this.barangayNote,
    required this.priorityEndorsed,
    this.priorityEndorsementReason,
    required this.attentionRequested,
    this.attentionRequestedReason,
    required this.recurringFlagged,
  });

  factory Report.fromJson(Map<String, dynamic> json) {
    return Report(
      id: json['id'] as String,
      title: json['title'] as String? ?? '',
      category: json['category'] as String?,
      subcategory: json['subcategory'] as String?,
      description: json['description'] as String? ?? '',
      address: json['address'] as String?,
      lat: (json['lat'] as num).toDouble(),
      lng: (json['lng'] as num).toDouble(),
      status: json['status'] as String? ?? 'Pending',
      severity: json['severity'] as String? ?? 'Moderate',
      severityConfidence: (json['severityConfidence'] as num?)?.toDouble(),
      severitySource: json['severitySource'] as String?,
      severityReasoning: json['severityReasoning'] as String?,
      similarReportCount: (json['similarReportCount'] as num?)?.toInt() ?? 1,
      submittedDate: DateTime.parse(json['submittedDate'] as String),
      photos: List<String>.from(json['photos'] ?? const []),
      resolutionPhotos: List<String>.from(json['resolutionPhotos'] ?? const []),
      barangayPhotos: List<String>.from(json['barangayPhotos'] ?? const []),
      history: (json['history'] as List? ?? const [])
          .map((e) => StatusEvent.fromJson(e as Map<String, dynamic>))
          .toList(),
      resolutionDescription: json['resolutionDescription'] as String?,
      resolutionDate: json['resolutionDate'] != null ? DateTime.tryParse(json['resolutionDate']) : null,
      barangay: json['barangay'] as String?,
      reporterName: json['reporterName'] as String?,
      invalidReason: json['invalidReason'] as String?,
      barangayConfirmed: json['barangayConfirmed'] as bool? ?? false,
      barangayNote: json['barangayNote'] as String?,
      priorityEndorsed: json['priorityEndorsed'] as bool? ?? false,
      priorityEndorsementReason: json['priorityEndorsementReason'] as String?,
      attentionRequested: json['attentionRequested'] as bool? ?? false,
      attentionRequestedReason: json['attentionRequestedReason'] as String?,
      recurringFlagged: json['recurringFlagged'] as bool? ?? false,
    );
  }
}

class StatusEvent {
  final String status;
  final DateTime date;
  final String? note;

  StatusEvent({required this.status, required this.date, this.note});

  factory StatusEvent.fromJson(Map<String, dynamic> json) {
    return StatusEvent(
      status: json['status'] as String,
      date: DateTime.parse(json['date'] as String),
      note: json['note'] as String?,
    );
  }
}

File: bantai-backend/app/Models/Report.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Builder;

class Report extends Model
{
    protected $keyType = 'string';
    public $incrementing = false;

    protected $fillable = [
        'id', 'client_uuid', 'user_id', 'lgu_id', 'barangay_id', 'category_id', 'subcategory_id',
        'title', 'description', 'address', 'lat', 'lng', 'status',
        'severity', 'severity_confidence', 'severity_source', 'severity_reasoning', 'severity_overridden_by',
        'cluster_id', 'similar_report_count',
        'verified_at', 'verified_by', 'resolved_at', 'resolved_by', 'closed_at', 'closed_by', 'cancelled_at',
        'resolution_description', 'resolution_date', 'invalid_reason',
        'barangay_confirmed', 'barangay_confirmed_at', 'barangay_confirmed_by', 'barangay_note',
        'priority_endorsed', 'priority_endorsement_reason', 'priority_endorsed_by',
        'attention_requested', 'attention_requested_reason', 'attention_requested_by',
        'recurring_flagged', 'recurring_flagged_by',
        'device_submitted_at',
    ];

    protected function casts(): array
    {
        return [
            'lat' => 'float',
            'lng' => 'float',
            'severity_confidence' => 'float',
            'barangay_confirmed' => 'boolean',
            'priority_endorsed' => 'boolean',
            'attention_requested' => 'boolean',
            'recurring_flagged' => 'boolean',
            'verified_at' => 'datetime',
            'resolved_at' => 'datetime',
            'closed_at' => 'datetime',
            'cancelled_at' => 'datetime',
            'resolution_date' => 'datetime',
            'barangay_confirmed_at' => 'datetime',
            'device_submitted_at' => 'datetime',
        ];
    }

    // -- Relationships ---------------------------------------------------
    public function user() { return $this->belongsTo(User::class); }
    public function lgu() { return $this->belongsTo(Lgu::class); }
    public function barangay() { return $this->belongsTo(Barangay::class); }
    public function category() { return $this->belongsTo(Category::class); }
    public function subcategory() { return $this->belongsTo(Subcategory::class); }
    public function cluster() { return $this->belongsTo(IssueCluster::class, 'cluster_id'); }
    public function photos() { return $this->hasMany(ReportPhoto::class); }
    public function statusHistory() { return $this->hasMany(StatusHistory::class)->orderBy('created_at'); }
    public function followers() { return $this->belongsToMany(User::class, 'report_followers'); }

    public function evidencePhotos()
    {
        return $this->photos()->where('kind', 'evidence');
    }

    public function resolutionPhotos()
    {
        return $this->photos()->whereIn('kind', ['resolution_before', 'resolution_after', 'resolution_document']);
    }

    public function barangayPhotos()
    {
        return $this->photos()->where('kind', 'barangay_evidence');
    }

    // -- Jurisdiction scope ------------------------------------------------
    // Role determines what a user may DO; jurisdiction determines what data
    // they may SEE. This mirrors the Node backend's canAccessReport(): a
    // System Admin sees everything, a citizen only their own, a Barangay
    // Administrator only their barangay, and LGU staff only their LGU.
    public function scopeVisibleTo(Builder $query, User $user): Builder
    {
        return match ($user->role) {
            'system_admin' => $query,
            'citizen' => $query->where('user_id', $user->id),
            'barangay_admin' => $query->where('barangay_id', $user->barangay_id),
            'lgu_official', 'lgu_admin' => $query->where('lgu_id', $user->lgu_id),
            default => $query->whereRaw('1 = 0'),
        };
    }
}
